common/src/main/java/com/wynntils/models/abilitytree/parser/AbilityTreeParser.java

```diff
@@ -0,0 +1,188 @@
+        StyledText nameStyledText = StyledText.fromComponent(itemStack.getHoverName());
+
+        AbilityTreeNodeState state = AbilityTreeNodeState.LOCKED;
+        StyledText actualName;
+        if (nameStyledText.getPartCount() == 1) {
+            actualName = nameStyledText;
+        } else {
+            actualName = nameStyledText.iterate((part, changes) -> {
+                // The part which is bolded is the actual name of the ability
+                if (!part.getPartStyle().isBold()) {
+                    changes.clear();
+                }
+
+                return IterationDecision.CONTINUE;
+            });
+            state = AbilityTreeNodeState.UNLOCKABLE;
+        }
+
+        List<StyledText> loreStyledText = LoreUtils.getLoreStyledText(itemStack);
+
+        int cost = 0;
+        List<String> blocks = new ArrayList<>();
+        String requiredAbility = null;
+        ArchetypeRequirement requiredArchetype = null;
+        String archetype = null;
+
+        List<StyledText> includedLines = new ArrayList<>(loreStyledText);
+
+        for (StyledText text : loreStyledText) {
+            Matcher matcher = text.getMatcher(NODE_POINT_COST_PATTERN);
+            if (matcher.matches()) {
+                cost = Integer.parseInt(matcher.group(1));
+                includedLines.remove(text); // skip in description
+                continue;
+            }
+
+            matcher = text.getMatcher(NODE_BLOCKS_ABILITY_PATTERN);
+            if (matcher.matches()) {
+                blocks.add(matcher.group(1));
+                continue;
+            }
+
+            matcher = text.getMatcher(NODE_REQUIRED_ABILITY_PATTERN);
+            if (matcher.matches()) {
+                requiredAbility = matcher.group(1);
+                includedLines.remove(text); // skip in description
+                continue;
+            }
+
+            matcher = text.getMatcher(NODE_REQUIRED_ARCHETYPE_PATTERN);
+            if (matcher.matches()) {
+                requiredArchetype = new ArchetypeRequirement(matcher.group(1), Integer.parseInt(matcher.group(3)));
+                includedLines.remove(text); // skip in description
+                continue;
+            }
+
+            matcher = text.getMatcher(NODE_ARCHETYPE_PATTERN);
+            if (matcher.matches()) {
+                archetype = matcher.group(1);
+                continue;
+            }
+
+            matcher = text.getMatcher(NODE_BLOCKED);
+            if (matcher.matches()) {
+                state = AbilityTreeNodeState.BLOCKED;
+                continue;
+            }
+
+            matcher = text.getMatcher(NODE_UNLOCKED);
+            if (matcher.matches()) {
+                state = AbilityTreeNodeState.UNLOCKED;
+                continue;
+            }
+        }
+
+        if (state == AbilityTreeNodeState.UNLOCKABLE || state == AbilityTreeNodeState.UNLOCKED) {
+            // Skip empty line + "click here to unlock" / "unlocked already"
```

Comment:
Are you sure these are always the two last lines?

Reply:
Yes, otherwise we wouldnt parse the unlockabel/unlocked state. I know it seems weird, but it should work, and this is only needed for data dumping.